crates/block-producer/src/batch_builder/batch.rs

```diff
@@ -31,12 +31,39 @@ pub type BatchId = Blake3Digest<32>;
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct TransactionBatch {
     id: BatchId,
-    updated_accounts: Vec<(TransactionId, TxAccountUpdate)>,
+    updated_accounts: BTreeMap<AccountId, AccountUpdate>,
     input_notes: Vec<InputNoteCommitment>,
     output_notes_smt: BatchNoteTree,
     output_notes: Vec<OutputNote>,
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct AccountUpdate {
+    init_state: Digest,
+    final_state: Digest,
+    transactions: Vec<TransactionId>,
+    details: AccountUpdateDetails,
+}
+
+impl AccountUpdate {
+    fn new(tx: &ProvenTransaction) -> Self {
+        Self {
+            init_state: tx.account_update().init_state_hash(),
+            final_state: tx.account_update().final_state_hash(),
+            transactions: vec![tx.id()],
+            details: tx.account_update().details().clone(),
+        }
+    }
+
+    /// Merges the transaction's update into this account update.
+    fn merge_tx(&mut self, tx: &ProvenTransaction) {
+        self.final_state = tx.account_update().final_state_hash();
+        self.transactions.push(tx.id());
```

Comment:
I would probably add an `assert!` somewhere here to make sure that `self.final_state == tx.account_update().init_state_hash()`.

Reply:
Which immediately broken some tests 😬 (`test_build_batch_success`) as it built a batch out of identical transactions.

I "fixed" by making the account IDs unique per tx in a batch, however probably would be a good idea to support building consecutive transactions on the same account. This would make test setup easier.